### PR TITLE
Spark 3.3: SQL Extensions for DROP TAG

### PR DESCRIPTION
**Pull request by hililiwei**
Co-authored-by: Amogh Jahagirdar <jahamogh@amazon.com>
Co-authored-by: chidayong <247070443@qq.com>

### CODE REVIEW

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java [hunk 1]
```diff
@@ -264,6 +264,54 @@ public void testCreateOrReplace() throws NoSuchTableException {
     Assert.assertEquals(first, ref.snapshotId());
   }
 
+  @Test
```
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java:267]: I think we should also have a test to ensure that DROP TAG does not work when the ref is a branch. If I missed the opposite case for DROP BRANCH, I think let's add that as well for completeness.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java [hunk 2]
```diff
@@ -264,6 +264,54 @@ public void testCreateOrReplace() throws NoSuchTableException {
...
+    sql("ALTER TABLE %s DROP TAG %s", tableName, tagName);
+    table.refresh();
+    ref = table.refs().get(tagName);
+    Assert.assertNull(ref);
+
+    AssertHelpers.assertThrows(
+        "Non-conforming tag name",
+        IcebergParseException.class,
+        "mismatched input '123'",
+        () -> sql("ALTER TABLE %s DROP TAG %s", tableName, "123"));
```
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java:285]: I think it's fine to include this test but could we separate it and have this test just focused on the happy case?

**Review by amogh-jahagirdar — COMMENTED** [review]
Mostly looks good to me thanks @hililiwei ! Just some test cases I think we can clean up

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java (hunk 2 above)
Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java:285]: +1

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java [hunk 3]
```diff
@@ -264,6 +264,54 @@ public void testCreateOrReplace() throws NoSuchTableException {
...
+    Table table = validationCatalog.loadTable(tableIdent);
+    String tagName = "t1";
+    table.manageSnapshots().createTag(tagName, table.currentSnapshot().snapshotId()).commit();
+    SnapshotRef ref = table.refs().get(tagName);
+    Assert.assertEquals(table.currentSnapshot().snapshotId(), ref.snapshotId());
+
+    sql("ALTER TABLE %s DROP TAG %s", tableName, tagName);
+    table.refresh();
+    ref = table.refs().get(tagName);
+    Assert.assertNull(ref);
```
Comment (jackye1995) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestTagDDL.java:279]: nit: we should prefer to add a description for all the assertions, there are many `assertNull` here that lack description

**Comment by jackye1995** [comment]
Overall looks good to me, just some nit comments about the tests

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 [hunk 4]
```diff
@@ -76,6 +76,7 @@ statement
     | ALTER TABLE multipartIdentifier createReplaceBranchClause                             #createOrReplaceBranch
     | ALTER TABLE multipartIdentifier createReplaceTagClause                                #createOrReplaceTag
     | ALTER TABLE multipartIdentifier DROP BRANCH (IF EXISTS)? identifier                   #dropBranch
+    | ALTER TABLE multipartIdentifier DROP TAG (IF EXISTS)? identifier                      #dropTag
```
[the remark this answers is not in the document]
Comment (jackye1995) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:79]: I think we discussed this in https://github.com/apache/iceberg/pull/6637#discussion_r1083330167, and the conclusion was that it makes the logic a bit more consistent to have them separated. Let me know what you think!

**Comment by jackye1995** [comment]
Restart CI

**Review by jackye1995 — APPROVED** [review]
looks good to me!

**Review by yyanyy — APPROVED** [review]
lgtm!

**Review by amogh-jahagirdar — APPROVED** [review]
Looks good to me! Thanks @hililiwei for the contribution!

File: spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4 (hunk 4 above)
[the remark this answers is not in the document]
Comment (singhpk234) [reply spark/v3.3/spark-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:79]: I see, was mostly coming from spark code base, considering this example : 

https://github.com/apache/spark/blob/46a234125d3f125ba1f9ccd6af0ec1ba61016c1e/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4#L125

If we have reached the concencus, then it should be fine.

**Review by singhpk234 — APPROVED** [review]
LGTM, Thanks @hililiwei !

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestBranchDDL.java [hunk 5]
```diff
@@ -238,6 +238,28 @@ public void testDropBranchDoesNotExist() {
         () -> sql("ALTER TABLE %s DROP BRANCH %s", tableName, "nonExistingBranch"));
   }
 
+  @Test
+  public void testDropBranchFailesForTag() throws NoSuchTableException {
```
Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestBranchDDL.java:242]: Just noticed typo: Should be `testDropBranchFailsForTag`

Comment (amogh-jahagirdar) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestBranchDDL.java:242]:
```suggestion
  public void testDropBranchFailsForTag() throws NoSuchTableException {
```

**Comment by jackye1995** [comment]
Looks like we have all comments addressed, I will go ahead to merge the PR. Thanks for the contribution @hililiwei , and thanks for everyone's review!

**Comment by hililiwei** [comment]
Thank you all for review 😄

**Pull request merged**